### PR TITLE
🌱 Bump controller-tools to v0.12.1

### LINKED ISSUE #1757: `verify-crds` test is flaky
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

`hack/verify-crds.sh`, which is used by `pull-cluster-api-provider-vsphere-verify-crds` prow job or `make verify-crds`, is flaky. The script occasionally returns errors, but it will run successfully when retried. For example, please see [this PR](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1739#issuecomment-1407443719).

**What did you expect to happen:**

Even if the script is executed several times, the results are the same.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

An example of a failed test: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1739/pull-cluster-api-provider-vsphere-verify-crds/1619032549575626752

**Environment:**

- Cluster-api-provider-vsphere version: I confirmed this issue on 4aafab65d5cfcde29026d01b70ca3065124cd019
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #1928: make verify-crds intermittently produces different output
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

If you run `make verify-crds` repeatedly, about 20% of the time, the generated CRDs are actually different, usually adding "cluster-api" as a category or setting an atomic map type.

**What did you expect to happen:**

Consistent CRD generation

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
Bumps controller-gen to fix the non-deterministic YAML generation.

Thx very much to @ntoofu for fixing the issue in controller-tools (https://github.com/kubernetes-sigs/controller-tools/pull/792)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1757
Fixes #1928

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by sbueringer** [comment]
I think we should consider cherry-picking this into all branches

File: config/default/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachinetemplates.yaml [hunk 1]
```diff
@@ -832,10 +832,10 @@ spec:
                                           being referenced
                                         type: string
                                     required:
```
Comment (sbueringer) [inline config/default/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachinetemplates.yaml:834]: These changes are expected:
```go
// TypedLocalObjectReference contains enough information to let you locate the
// typed referenced object inside the same namespace.
// +structType=atomic
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

**Comment by chrischdi** [comment]
/lgtm

Did re-run the script successfully with 200 times no diff :-) 🎉 We're back to deterministic output 😄 

```
cd hack/tools
rm bin/controller-gen
make controller-gen
cd ../..
function test:generate() {
  rm -rf tmp tmp2
  mkdir -p tmp tmp2
  eval $CONTROLLER_GEN ${1}
  for i in {1..100}; do
    printf "> %02d\n" "$i";
    eval $CONTROLLER_GEN ${1}2
    diff tmp${2} tmp2${2} || break
  done
}

CONTROLLER_GEN="$(pwd)/hack/tools/bin/controller-gen"
GENARGS="paths=./apis/v1alpha3 \
    paths=./apis/v1alpha4 \
    paths=./apis/v1beta1 \
    crd:crdVersions=v1 \
    output:crd:dir=tmp"

test:generate ${GENARGS}
```

**Comment by sbueringer** [comment]
Thx for verifying and thx for investigating this issue!

**Pull request merged**